src/rules/noRestrictedGlobalsRule.ts

```diff
@@ -0,0 +1,122 @@
+            \`\`\`
+        `,
+        descriptionDetails: Lint.Utils.dedent`
+            Disallowing usage of specific global variables can be useful if you want to allow
+            a set of global variables by enabling an environment, but still want to disallow
+            some of those.
+        `,
+        optionsDescription: Lint.Utils.dedent`
+            This rule takes a list of strings, where each string is a global to be restricted.
+            \`event\`, \`name\` and \`length\` are restricted by default.
+        `,
+        options: {
+            type: "list",
+            items: { type: "string" },
+        },
+        optionExamples: [
+            [
+                true,
+                "name",
+                "length",
+                "event",
+            ],
+        ],
+        type: "functionality",
+        typescriptOnly: false,
+        requiresTypeInfo: true,
+    };
+    /* tslint:enable:object-literal-sort-keys */
+
+    public static FAILURE_STRING(name: string) {
+        return `Unexpected global variable '${name}'. Use a local parameter or variable instead.`;
+    }
+
+    public applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program): Lint.RuleFailure[] {
+        const bannedList = this.ruleArguments.length > 0 ? this.ruleArguments : ["event", "name", "length"];
+        const bannedGlobals = new Set(bannedList);
+        if (sourceFile.isDeclarationFile) {
+            return [];
+        } else {
+            return this.applyWithFunction(sourceFile, walk, bannedGlobals, program.getTypeChecker());
+        }
+    }
+}
+
+function walk(ctx: Lint.WalkContext<Set<string>>, checker: ts.TypeChecker): void {
+    return ts.forEachChild(ctx.sourceFile, function recur(node: ts.Node | undefined): void {
+        if (node == undefined) {
+            return;
+        }
+
+        // Handles `const { bar, length: { x: y = () => event } } = foo`
+        if (isBindingElement(node)) {
+            recur(node.initializer);
+            recur(node.name);
+            if (node.propertyName != undefined && isComputedPropertyName(node.propertyName)) {
+                recur(node.propertyName);
+            }
+        } else if (isPropertyAccessExpression(node)) {
+            // Ignore `y` in `x.y`, but recurse to `x`.
+            recur(node.expression);
+        } else if (isIdentifier(node)) {
+            checkIdentifier(node);
+        } else {
+            ts.forEachChild(node, recur);
+        }
+    });
+
+    function checkIdentifier(node: ts.Identifier): void {
+        if (!ctx.options.has(node.text)) {
+            return;
+        }
+
+        const symbol = checker.getSymbolAtLocation(node);
+        const declarations = symbol === undefined ? undefined : symbol.declarations;
+        if (declarations === undefined || declarations.length === 0) {
+            return;
+        }
+
+        const isAmbientGlobal = declarations.some((decl) => decl.getSourceFile().isDeclarationFile);
```

Comment:
This errors too often if the declaration comes from a declaration file that is *not* lib.dom.d.ts. Consider the case of adone's types in Definitely Typed:

```ts
// @Filename: adone-tests.ts
namespace eventsTests {
    namespace EventEmitter {
        namespace static {
            const a: number = adone.event.Emitter.listenerCount(new adone.event.Emitter(), "event");
            const b: number = adone.event.Emitter.defaultMaxListeners;
        }
// @Filename: events.d.ts
declare namespace adone {
    namespace event {
        class Emitter {
            static listenerCount(emitter: Emitter, event: string | symbol): number;
            static defaultMaxListeners: number;
            // ...
```

You will get bogus errors on `event`.